File: app/src/main/java/com/digzdigital/divinitytoday/ui/devlist/adapter/DevotionalAdapter.java
package com.digzdigital.divinitytoday.ui.devlist.adapter;

import android.support.v7.widget.CardView;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.digzdigital.divinitytoday.R;
import com.digzdigital.divinitytoday.data.Devotional;

import java.util.ArrayList;


public class DevotionalAdapter extends RecyclerView.Adapter<DevotionalAdapter.ViewHolder> {

    private static MyClickListener myClickListener;
    private static final int VIEW_TYPE_ITEM = 0;
    private static final int VIEW_TYPE_LOADING = 1;
    private ArrayList<Devotional> devotionals;


    public DevotionalAdapter(ArrayList<Devotional> devotionals) {
        this.devotionals = devotionals;
    }


    @Override
    public int getItemViewType(int position) {
        return devotionals.get(position) == null ? VIEW_TYPE_LOADING : VIEW_TYPE_ITEM;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.devotionals_grid, parent, false);
            return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
            Devotional devotional = getItem(position);
            holder.devTitle.setText(devotional.getTitle());
            holder.devDate.setText(devotional.getDate());
    }

    @Override
    public int getItemCount() {
        return devotionals == null ? 0 : devotionals.size();
    }

    public void setOnItemClickListener(MyClickListener myClickListener) {
        DevotionalAdapter.myClickListener = myClickListener;
    }

    private Devotional getItem(int position) {
        return devotionals.get(position);
    }

    public interface MyClickListener {
        void onItemClick(int position, View v);
    }


    static class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        CardView cv;
        TextView devTitle, devDate;

        ViewHolder(View itemView) {
            super(itemView);
            cv = (CardView) itemView.findViewById(R.id.cv);
            devTitle = (TextView) itemView.findViewById(R.id.dev_title);
            devDate = (TextView) itemView.findViewById(R.id.dev_date);
            itemView.setOnClickListener(this);
        }


        @Override
        public void onClick(View v) {
            myClickListener.onItemClick(getAdapterPosition(), v);

        }

    }

}

File: app/src/main/java/com/digzdigital/divinitytoday/ui/devlist/adapter/DevotionalAdapter.kt
package com.digzdigital.divinitytoday.ui.devlist.adapter

import android.support.v4.util.SparseArrayCompat
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.ViewGroup
import com.digzdigital.divinitytoday.commons.MyClickListener
import com.digzdigital.divinitytoday.commons.adapter.AdapterConstants
import com.digzdigital.divinitytoday.commons.adapter.ViewType
import com.digzdigital.divinitytoday.commons.adapter.ViewTypeDelegateAdapter
import com.digzdigital.divinitytoday.data.model.Devotional
import com.digzdigital.divinitytoday.data.model.DevotionalAd

class DevotionalAdapter(myClickListener: MyClickListener) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    var items: ArrayList<ViewType>
    private var delegateAdapters = SparseArrayCompat<ViewTypeDelegateAdapter>()
    private val loadingItem = object : ViewType {
        override fun getViewType() = AdapterConstants.LOADING
    }

    init {
        delegateAdapters.put(AdapterConstants.LOADING, LoadingDelegateAdapter())
        delegateAdapters.put(AdapterConstants.DEVOTIONAL, DevotionalDelegateAdapter(myClickListener))
        delegateAdapters.put(AdapterConstants.AD, AdDelegateAdapter())
        items = java.util.ArrayList()
        items.add(loadingItem)
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        delegateAdapters.get(getItemViewType(position)).onBindViewHolder(holder, this.items[position])
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return delegateAdapters.get(viewType).onCreateViewHolder(parent)
    }

    override fun getItemCount(): Int {
        return items.size
    }

    override fun getItemViewType(position: Int): Int {
        return this.items[position].getViewType()
    }

    fun addDevotionals(news: List<Devotional>) {
        val initPosition = items.size - 1
        items.removeAt(initPosition)
        notifyItemRemoved(initPosition)

        items.addAll(news)
        items.add(loadingItem)
        notifyItemRangeChanged(initPosition, items.size + 1)
    }

    fun addAds(position:Int, devotionalAd: DevotionalAd){
        items.add(position, devotionalAd)
        notifyItemInserted(position)
    }

    fun clearAndAddDevotionals(news: List<Devotional>) {
        val lastPosition = getLastPosition()
        items.clear()
        notifyItemRangeRemoved(0, lastPosition)

        items.addAll(news)
        items.add(loadingItem)
        notifyItemRangeInserted(0, items.size)
        notifyDataSetChanged()
    }

    fun getDevotionals(): List<Devotional> {
        return items
                .filter { it.getViewType() == AdapterConstants.DEVOTIONAL }
                .map { it as Devotional }
    }

    private fun getLastPosition() = if (items.lastIndex == -1) 0 else items.lastIndex


}